File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    ans = 0
    visited = [False] * n

    def bfs(start):
        que = deque([start])
        visited[start] = True
        
        while que:
            q = que.popleft()
            for i in range(n):
                if not visited[i] and computers[q][i]:
                    visited[i] = True
                    que.append(i)

    for i in range(n):
        if not visited[i]:
            bfs(i)
            ans += 1

    return ans

File: 프로그래머스/lv3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    let ans = 0;
    
    const visited = new Array(n).fill(false); // 방문 처리 배열
    
    const BFS = (start) => {
        const que = [start];
        visited[start] = true;
        
        while(que.length > 0){
            const q = que.shift();
            computers[q].forEach((value, index) => {
                if(!visited[index] && value === 1){ // 방문 하지 않았고 연결되어 있다면
                    que.push(index);
                    visited[index] = true; // 방문처리
                }
            })
        }
        return 1; // 네트워크 + 1
    }
    
    for(const computer of computers){
        computer.forEach((value, index) => {
            if(!visited[index] && value === 1){
                // 방문하지 않았고 다른 컴퓨터와 연결되어 있다면
                ans += BFS(index);
            }
        });
    }
    
    return ans;
}